File: general_algorithm/greedy/0435_non_overlapping_intervals.js
/*
Given an array of intervals intervals where intervals[i] = [starti, endi], 
return the minimum number of intervals you need to remove to make the rest 
of the intervals non-overlapping.

1 <= intervals.length <= 10^5
intervals[i].length == 2
-5 * 10^4 <= starti < endi <= 5 * 10^4
*/

// https://leetcode.com/problems/non-overlapping-intervals/description/

/* since we want to remove the minimum number of intervals, if two intervals 
overlap, we want to remove the larger one.
sort the intervals by the start, then loop through the intervals.
*/
var eraseOverlapIntervals = function (intervals) {
  // sort the intervals by the start
  intervals.sort((a, b) => a[0] - b[0]);
  let count = 0,
    range = [];
  // loop through the intervals
  for (let [start, end] of intervals) {
    // initialize range with the first interval
    if (!range.length) {
      range = [start, end];
      continue;
    }
    // if the current interval overlaps, remove the larger interval
    if (start < range[1]) {
      count++;
      // remove the larger interval
      range[1] = Math.min(range[1], end);
    }
    // if the current interval does not overlap with last interval
    // update range to the current interval
    else {
      range = [start, end];
    }
  }
  return count;
};

File: general_algorithm/greedy/0435_non_overlapping_intervals.py
# Given an array of intervals intervals where intervals[i] = [starti, endi], return the minimum number of intervals you need to remove to make the rest of the intervals non-overlapping.

# Note that intervals which only touch at a point are non-overlapping. For example, [1, 2] and [2, 3] are non-overlapping.

# https://leetcode.com/problems/non-overlapping-intervals/

# greedy
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0
        # Sort intervals by their end times
        intervals.sort(key=lambda interval: interval[1])
        
        # Initialize with the end time of the first interval
        end = intervals[0][1]
        remove_count = 0
        
        # Iterate through intervals starting from the second one
        for i in range(1, len(intervals)):
            if intervals[i][0] < end:
                # Overlap found, increment remove count
                remove_count += 1
            else:
                # No overlap, update end to the current interval's end
                end = intervals[i][1]
        
        return remove_count